src/main/java/com/palantir/common/streams/BackpressureSpliterator.java

```diff
@@ -0,0 +1,115 @@
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+class BackpressureSpliterator<U, V> implements Spliterator<V> {
+    private final int desiredConcurrency;
+    private final BlockingQueue<CompletableFuture<V>> completed;
+    private final Spliterator<U> notStarted;
+    private final Function<U, CompletableFuture<V>> mapper;
+
+    private int inProgress = 0;
+
+    private BackpressureSpliterator(
+            int desiredConcurrency,
+            Spliterator<U> arguments,
+            Function<U, CompletableFuture<V>> mapper) {
+        checkArgument(desiredConcurrency > 0, "desiredConcurrency %s > 0", new Object[] { desiredConcurrency });
+        this.desiredConcurrency = desiredConcurrency;
+        this.completed = new ArrayBlockingQueue<>(desiredConcurrency);
+        this.notStarted = arguments;
+        this.mapper = mapper;
+    }
+
+    @Override
+    public boolean tryAdvance(Consumer<? super V> action) {
+        startNewWorkIfNecessary();
+        if (inProgress == 0) {
+            return false;
+        }
+
+        try {
+            V element = completed.take().join();
+            inProgress--;
+            action.accept(element);
+            return true;
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public Spliterator<V> trySplit() {
+        return null;
+    }
+
+    private void startNewWorkIfNecessary() {
+        while (inProgress < desiredConcurrency) {
+            boolean maybeRemainingElements = notStarted.tryAdvance(nextInput -> {
+                CompletableFuture<V> nextFuture = mapper.apply(nextInput);
+                inProgress++;
+                nextFuture.whenComplete((res, err) -> completed.add(nextFuture));
+            });
+            if (!maybeRemainingElements) {
+                return;
+            }
+        }
+    }
+
+    @Override
+    public long estimateSize() {
+        long estimate = completed.size() + inProgress + notStarted.estimateSize();
+        if (estimate < 0L) {
+            return Long.MAX_VALUE;
+        }
+        return estimate;
+    }
+
+    @Override
+    public long getExactSizeIfKnown() {
+        long delegateResult = notStarted.getExactSizeIfKnown();
+        if (delegateResult == -1L) {
+            return -1L;
+        }
+        return completed.size() + inProgress + delegateResult;
+    }
+
+    @Override
+    public int characteristics() {
+        return (Spliterator.SIZED | Spliterator.SUBSIZED) & notStarted.characteristics();
+    }
+
+    static <U, V> Spliterator<V> create(
+            int desiredConcurrency,
+            Stream<U> arguments,
+            Function<U, CompletableFuture<V>> mapper) {
+        BackpressureSpliterator<U, V> result = new BackpressureSpliterator<>(
+                desiredConcurrency, arguments.spliterator(), mapper);
+        result.startNewWorkIfNecessary();
```

Comment:
Don't do this here

Reply:
As in, better to let the caller kick it off when they start to consume the output